lib/shoulda/matchers/active_model/comparison_matcher.rb

```diff
@@ -0,0 +1,158 @@
+          },
+        }.freeze
+
+        delegate :failure_message, :failure_message_when_negated, to: :submatchers
+
+        def initialize(matcher, value, operator)
+          super(nil)
+          unless matcher.respond_to? :diff_to_compare
+            raise ArgumentError, 'matcher is invalid'
+          end
+
+          @matcher = matcher
+          @value = value
+          @operator = operator
+          @message = ERROR_MESSAGES[operator][:label]
+        end
+
+        def simple_description
+          description = ''
+
+          if expects_strict?
+            description << ' strictly'
+          end
+
+          description +
+            "disallow :#{attribute} from being a number that is not " +
+            "#{comparison_expectation} #{@value}"
+        end
+
+        def for(attribute)
+          @attribute = attribute
+          self
+        end
+
+        def with_message(message)
+          @expects_custom_validation_message = true
+          @message = message
+          self
+        end
+
+        def expects_custom_validation_message?
+          @expects_custom_validation_message
+        end
+
+        def matches?(subject)
+          @subject = subject
+          build_option_value
+          submatchers.matches?(subject)
+        end
+
+        def does_not_match?(subject)
+          @subject = subject
+          build_option_value
+          submatchers.does_not_match?(subject)
+        end
+
+        def comparison_description
+          "#{comparison_expectation} #{@value}"
+        end
+
+        def submatchers
+          @_submatchers ||= NumericalityMatchers::Submatchers.new(build_submatchers)
+        end
+
+        private
+
+        def build_submatchers
+          comparison_combos.map do |diff, submatcher_method_name|
+            matcher = __send__(submatcher_method_name, diff, nil)
+            matcher.with_message(@message, values: { count: @value })
+            matcher
+          end
+        end
+
+        def comparison_combos
+          diffs_to_compare.zip(submatcher_method_names)
+        end
+
+        def submatcher_method_names
+          assertions.map do |value|
+            if value
+              :allow_value_matcher
+            else
+              :disallow_value_matcher
+            end
+          end
+        end
+
+        def assertions
+          ERROR_MESSAGES[@operator][:assertions]
+        end
+
+        def build_option_value
+          @value = case @value
+                   when Proc then @value.call(@subject)
+                   when Symbol then @subject.send(@value)
+                   else @value
+                   end
+        end
+
+        def diffs_to_compare
+          diff_to_compare = @matcher.diff_to_compare
+          values = case @value
+                   when String then diffs_when_string(diff_to_compare)
+                   else [-1, 0, 1].map { |sign| @value + (diff_to_compare * sign) }
+                   end
```

Comment:
This change was required to make a comparison between strings possible, as we can see on the specs for the comparison implementation in Rails.

https://github.com/ChaelCodes/rails/blob/2ed77a9577d75579eabda04f8abb3a18c71322ea/activemodel/test/cases/validations/comparison_validation_test.rb#L33